include/libcyphal/executor.hpp

```diff
@@ -219,64 +125,30 @@ class IExecutor : public cetl::rtti_helper<IExecutorTypeIdType>
-    /// @return A new unique identifier for the callback if successful.
-    ///         Otherwise `nullopt` in case of out of memory error.
-    ///         No discard b/c it's expected to be used in conjunction with `scheduleCallbackByIdAt`
-    ///         or `removeCallbackById` methods.
+    /// @param function The function to be called when the callback is executed.
+    /// @return Type-erased instance of the registered callback. Instance must not outlive the executor,
+    ///         and must be used only with the same executor; otherwise undefined behavior.
     ///
-    CETL_NODISCARD virtual cetl::optional<Callback::Id> appendCallback(Callback::Function&& function) = 0;
+    CETL_NODISCARD virtual Callback::Any registerCallback(Callback::Function&& function) = 0;
 
-    /// @brief Schedules previously appended callback (by its id) for execution at the desired absolute time.
+    /// @brief Schedules previously registered callback for execution according to a desired schedule.
     ///
     /// Actual execution of the callback's function will be done later (not from context of this method), when desired
-    /// time comes and executor is ready to execute the callbacks. It's ok to schedule the same callback multiple times
+    /// time comes and executor is ready to execute callbacks. It's ok to schedule the same callback multiple times
     /// even before it was executed - it will be rescheduled, and then executed according to the last setup.
     ///
-    /// @param callback_id Unique identifier of the callback to be scheduled.
-    /// @param time_point Absolute time point when it's desired to execute it.
-    ///                   Use current time (aka now) to schedule it for ASAP execution.
-    ///                   It could be in the past as well - callback will be executed as soon as possible.
-    /// @param schedule Contains specifics how callback will be scheduled (like once, repeatedly etc.).
-    /// @return `true` if the callback was found and successfully scheduled.
-    ///         Otherwise, in case the callback has been removed already (f.e. by auto removal on execution),
-    ///         or there are not enough resources to schedule the callback.
+    /// @param callback The callback instance (registered at this executor) to be scheduled.
+    ///                 There will be no scheduling if callback is not registered yet
+    ///                 (in its initial default value), or has been reset already.
+    /// @param schedule Contains specifics of how exactly callback will be scheduled (like once, repeatedly etc.).
     ///
-    virtual bool scheduleCallbackById(const Callback::Id                 callback_id,
-                                      const TimePoint                    exec_time,
-                                      const Callback::Schedule::Variant& schedule) = 0;
+    virtual void scheduleCallback(Callback::Any& callback, const Callback::Schedule::Variant& schedule) = 0;
```

Comment:
Plz see above docs for `callback` parameter - so, initial/default (or `.reset()`-ed) callback instance is the only way how it could kinda "fail". If you think it deserves `bool` (instead of `void`) let me know - initially I did it like this (with `bool`) but later I inclined to just document expectation.

For the "past" question please see documentation for Schedule. In short - it will be executed asap.

Reply:
Here's my concern:
```
MyObject* p = new MyObject();
e->scheduleCallback([p]{ delete p; }); <-- if this isn't scheduled then we have a memory leak.
```

Reply:
Should you in the first place use unique ptr for `MyObject`, and capture it by move? Also, even if you would have scheduling `true` result but later call `callback.reset()` (before the callback had a chance to be executed) then you will have the same "leak" (but not with smart ptr).
But in general I've got your concern. ok, I'll try with `bool` return and see how it will go inside libcyphal... but I anticipate complications and surplus of extra corner cases and failure points in originally quite safe places.

Honestly, I would rather have extra bunch of asserts which will fail if you try schedule already reset callback. And also make `.schedule` method as part of `Callback::Any` type (like `callback.shedule(...)` instead of current `executor.scheduleCallback(callback, ...)`) to eliminate what I've describe in docs as "_...must be used only with the same executor..._" - I discussed this move with @pavel-kirienko , and we agreed to do it (but in separate PR).

Reply:
Let's open an issue up on this and then proceed with this CR as-is. I'm concerned about the case where the scheduled time can become in the past between the time it is calculated and the time the schedule method is invoked.